campaignion_donation/components/donation_amount.php

```diff
@@ -0,0 +1,213 @@
+<?php
+
+/**
+ * @file
+ * Donation amount webform component.
+ */
+
+/**
+ * Implements _webform_defaults_component().
+ */
+function _webform_defaults_donation_amount() {
+  $defaults = webform_component_invoke('number', 'defaults');
+  unset($defaults['type']);
+  $defaults['extra'] += [
+    'options' => [],
+    'other_option' => TRUE,
+    'other_text' => t('Other'),
+  ];
+  return $defaults;
+}
+
+/**
+ * Fix the view field(s) that are automatically generated for number components.
+ */
+function _webform_view_field_donation_amounts($component, $fields) {
+  foreach ($fields as &$field) {
+    $field['webform_datatype'] = 'number';
+  }
+  return $fields;
+}
+
+/**
+ * Implements _webform_edit_component().
+ */
+function _webform_edit_donation_amount($component) {
+  $number_component = $component;
+  $component['extra']['type'] = 'textfield';
+  $form = webform_component_invoke('number', 'edit', $component);
+  unset($form['display']['type']);
+  $form['behavior'] = [
+    '#type' => 'fieldset',
+    '#title' => t('Behavior'),
+  ];
+  $form['behavior']['currency'] = [
+    '#type' => 'select',
+    '#title' => t('Currency'),
+    '#options' => webform_currency_options(),
+    '#default_value' => $component['extra']['currency'],
+    '#parents' => ['extra', 'currency'],
+  ];
+  $form['behavior']['options'] = [
+    '#type' => 'textarea',
+    '#title' => t('Pre-defined amounts'),
+    '#description' => t('Enter pre-defined donation amounts separated by a line break. Use “.” as decimal point.'),
+    '#default_value' => implode("\n", $component['extra']['options']),
+    '#parents' => ['extra', 'options'],
+    '#element_validate' => ['_webform_edit_donation_amount_validate_amounts'],
+  ];
+  $form['behavior']['other_option'] = [
+    '#type' => 'checkbox',
+    '#title' => t('Allow "Other..." option'),
+    '#default_value' => $component['extra']['other_option'],
+    '#description' => t('Check this option if you want to allow users to enter an option not on the list.'),
+    '#parents' => ['extra', 'other_option'],
+    '#attributes' => ['class' => ['other-option-checkbox']],
+  ];
+  $form['behavior']['other_text'] = [
+    '#type' => 'textfield',
+    '#title' => t('Text for "Other..." option'),
+    '#default_value' => $component['extra']['other_text'],
+    '#description' => t('If allowing other options, enter text to be used for other-enabling option.'),
+    '#parents' => ['extra', 'other_text'],
+    '#weight' => 3,
+    '#states' => [
+      'visible' => [
+        ':input.other-option-checkbox' => ['checked' => TRUE],
+      ],
+    ],
+  ];
+
+  return $form;
+}
+
+/**
+ * Element validate callback: Split amount array.
+ */
+function _webform_edit_donation_amount_validate_amounts($element, &$form_state) {
+  $values =
```

Comment:
This line does not seem to be terminated... (or I do not know the syntax/code style for this code...)

Reply:
But as `$values` is set further down below I guess this line can be removed...